### ISSUE TITLE
config-driven augmentations

### ISSUE THREAD

**Issue by eminorhan** [AI-HHMI/miao#32]
Augmentations are currently composed in caller code and passed to `VolumeDataset` as `augment_fn` (this is a change introduced in 9e79d38). This change removed a lot of unmaintainable complexity from the core code, but augmentations are no longer included in the configs. I'd like to discuss if this choice is ideal. One could make the argument that the configs should ideally include a full description of what is done to the data during training for purposes of reproducibility.

Here's a proposal for bringing back config-driven augmentations without incurring the complexity cost of the previous version (`aug_rot`).

The augmentations would be described as an ordered list of named operations in the config:

```yaml
augment:
  - rot90isocube:     {}
  - shift_sections:   {prob: 0.05, magnitude: 10, p: 0.5}
  - drop_sections:    {prob: 0.05, p: 0.5}
  - intensity_jitter: {p: 0.5}
```

- We would have a registry of augmentations, mapping `name -> (function, geometric?)`. `geometric` says whether the op is handed the labels alongside the image (it's a property of the function, not a config choice).

- We would have one validator for the whole registry. It checks each entry's name and keyword arguments against the function's own `inspect.signature`, so a typo fails at config load rather than at the first sample or silently.

- We would have one application path: `AugmentPipeline`, a picklable class that `VolumeDataset` builds from `config.augment`, forwarding `output_axes` and the per-sample `pixel_size` to any function that accepts them.

- Adding a new augmentation function becomes one registry line, after which it is configurable and argument-validated with no schema, validator or `VolumeDataset` change.

@jakobtroidl @colemanbroad Please let me know what you think about this proposal.

**Comment by jakobtroidl** [comment]
Yes, I think this proposal makes lots of sense! @colemanbroad wdyt ?

**Comment by colemanbroad** [comment]
This means our config language attempts to take on the full complexity of the imperative python code in `augment_fn`, but I suspect yaml will fall short. How would we translate the augment examples in the README, in particular a) mapping a rotation across levels b) applying an aug to both `img` and `lab`? I can imagine using the configs to define a restrictive subset of possible augmentations making reasonable assumptions for these and other choices, but it would be extra work to maintain when reproducibility is already achieved by reusing an existing `augment_fn`.

**Comment by eminorhan** [comment]
The config would be able to describe augmentations that can be expressed as a flat ordered list of an arbitrary number of basic augmentations (intended to be similar to how `torchvision.transforms.Compose()` handles transforms). I think this should be able to cover most, if not all, cases of practical interest.

I think the examples you mentioned in the README can also be handled naturally by the config-driven augmentations. You would only need to do something like this in the config:

```yaml
output_axes: "lzyxc"
resolutions: [[1,1,1], [2,2,2], [4,4,4]]

augment:
  - rot90isocube: {}
  - intensity_jitter: {}
```
and this would be able to handle both examples you mentioned:

b) The registry would record that `rot90isocube` is geometric and `intensity_jitter` is not. So the pipeline would pass both tensors to the first but only the image tensor to the second. This would be a fact about the function, so the config would never have to specify it.

a) Rotation across levels would also be automatic with something along the lines of:
```python
def spatial_dims(axes):
    return tuple(axes.index(a) - len(axes) for a in "zyx")

perm, flips = draw_rot90(rng)                                                     # ONE draw
img = apply_rot90(img, perm, flips, spatial_dims(output_axes))                    # (-4,-3,-2)
lab = apply_rot90(lab, perm, flips, spatial_dims(output_axes.replace("c","")))    # (-3,-2,-1)
```
using the exact same config above. The thing to remember here is that the config already knows the layout (in `output axes`). Another option is to allow a `spatial_dims` override in the config for greater flexibility. In any case, I don't think either of these examples poses any real difficulty for the config-driven design.  

My main concern with the current approach is that augmentations are a crucial part of a training pipeline. If this is not in the `miao` config, sharing the config between different people will not be enough for reproducibility if they're using different code bases to train their models. They would have to go to the code base to recover what augmentations were applied.

**Comment by colemanbroad** [comment]
I expect the set of augmentations and their api to continue to grow/churn significantly which means configs will often be stale. The `p: 0.5` arguments are control flow in disguise, better expressed as `if`. Eventually people will ask for `one_of` to express "one of these two augs, randomly" like [albumentations](https://albumentations.ai/docs/api-reference/albumentations/core/composition/#OneOf). In YAML we have no types, no debugger, no stack traces, no IDE completion, evaluated by a bespoke interpreter we now maintain. The right way to validate these pipelines is just to run them.

"Always apply the same rotation at every level" is an assumption we're forced to bake to avoid make the config lang even worse. Transformations like those in `labels.py` complicate the story further as they return new data, not just ops on `img` and `label`. That's handled naturally by `augment_fn`.

I am totally sympathetic to the desire to express the full training data pipeline in the config (really! 😄). How would we feel about something like

```python
augment_fn: myproject.augs.my_em_augfn   # dotted path, resolved with one importlib call
# could also be
augment_fn: miao.augs.em_default   # we provide a set of defaults
```

which has the added benefit that it's much easier to share a pipeline / check if two augmentation pipelines are the same.

**Comment by eminorhan** [comment]
```yaml
augment_fn: myproject.augs.my_em_augfn   # dotted path, resolved with one importlib call
# could also be
augment_fn: miao.augs.em_default   # we provide a set of defaults
```
Yes, I think this is better than nothing. I think this has its own problems, for example, it only records an abstract function name, but what if that function changes? Or what if I want the same augmentation pipeline as `myproject.augs.my_em_augfn` but with `scale=0.2` (the config handles this much more naturally)? But it's definitely better than nothing, which is the current state.

**Comment by eminorhan** [comment]
Something like this:
```yaml
augment_fn:
  factory: miao.augs.em_default
  kwargs: {scale: 0.2}
```
would address the second issue.

**Comment by jakobtroidl** [comment]
Closing for now as we aligned on the [above convention](https://github.com/AI-HHMI/miao/issues/32#issuecomment-5320488762).
